File: EPIJudge/epi_judge_cpp/sorted_arrays_merge.cc
#include <vector>
#include <bits/stdc++.h>
#include <iterator>     // std::next

#include "test_framework/generic_test.h"
using namespace std;

struct VectorIt {
    bool operator()(const VectorIt &p1, const VectorIt &p2) const {
        return *(p1.cur) > *(p2.cur);
    }

    vector<int>::const_iterator cur, end;
};

vector<int> MergeSortedArrays(const vector<vector<int>>& sorted_arrays) {
    priority_queue<VectorIt, vector<VectorIt>, VectorIt> pq;
    for (const auto& it: sorted_arrays) {
        if (!it.empty()) {
            pq.emplace(VectorIt{it.begin(), it.end()});
        }
    }

    vector<int> out;
    while (!pq.empty()) {
        VectorIt it = pq.top();
        pq.pop();
        out.emplace_back(*(it.cur));
        if (next(it.cur) != it.end) {
            pq.emplace(VectorIt{next(it.cur), it.end});
        }
    }
    return out;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"sorted_arrays"};
  return GenericTestMain(args, "sorted_arrays_merge.cc",
                         "sorted_arrays_merge.tsv", &MergeSortedArrays,
                         DefaultComparator{}, param_names);
}

File: EPIJudge/epi_judge_python/sorted_arrays_merge.py
from typing import List
import heapq
from test_framework import generic_test


def merge_sorted_arrays(sorted_arrays: List[List[int]]) -> List[int]:
    ind = [0] * len(sorted_arrays)
    q = []
    for i in range(len(sorted_arrays)):
        heapq.heappush(q, (sorted_arrays[i][0], i))

    ans = []
    while len(q):
        x, i = heapq.heappop(q)
        ans.append(x)
        ind[i] += 1
        if ind[i] < len(sorted_arrays[i]):
            heapq.heappush(q, (sorted_arrays[i][ind[i]], i))

    return ans



if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('sorted_arrays_merge.py',
                                       'sorted_arrays_merge.tsv',
                                       merge_sorted_arrays))
